### PR TITLE
Add VideoPlayer Extension to All Platforms

### PR DESCRIPTION
**Pull request by time-killer-games**
- Win32 games need MSYS2 installed along with the mpv/libmpv package to get all the deps.
- Mac users need to install mpv/libmpv via homebrew or similar package managers available.
- Linux and FreeBSD users may simply use their package manager as usual to install libmpv.
- Android support is also possible due libmpv has an Android build - so we have all platforms.

### CODE REVIEW

**Comment by fundies** [comment]
I don't think we should be adding an extensions to launch specific programs. I think if we want a video extension it should be rendering to a surface directly and not launching some video player. If you want a general extension to launch programs I'm fine with that. You could then provide all these video_launch functions as scripts in a game instead. That way when you want to launch a dialog, or a video player or whatever it all uses the same code.

**Comment by time-killer-games** [comment]
@fundies Josh said he doesn't want there to be an extension which does stuff with application process id's exposed to the user namespace and that I should host it on my own repository if people really want process stuff. My insecure_bullshit.cpp's would go better in the engine for internal use instead. It still won't be exposed to enigma_user. That way if multiple extensions need these features they'll have access to it in a place where it makes more sense than just in my video player extension lol.

The video embeds in the game window, so at least it's similar to a surface in that it doesn't open in it's own window. Similar to zenity/kdialog, doing it from a command line program bypasses the need to distribute your game's source because you aren't linking to it in the form of a static or dynamic library. I'm not completely against rethinking whether this extension should be in enigma, but I do know hugar seemed to get really excited about it, I want Josh's and Robert's opinion before I call it quits.

**Comment by fundies** [comment]
I don't particularly want that extension either. I was just saying I'd prefer you'd reuse that functionality if you're going to insist on it. Anyways, just displaying a video in the window is no good. A proper video extension should be able to render to a texture in game for stuff like displaying a video on a TV asset in the game but this will require direct usage of a c++ or c library.

**Comment by time-killer-games** [comment]
I think personally this asset is pretty decent for what most real games need videos for; mainly animated logos and brief cut-scenes which pretty much always cover the entire screen, but that's just my opinion. Most games really don't need anything that fancy. It would also be pretty simple to add a custom pause feature but I haven't gotten around to that just yet. I've rarely seen a video rendered to a surface in games I've seen use videos.

**Comment by time-killer-games** [comment]
@fundies I think you will like this, I have written a video player using DirectShow with some help from someone from the BlitzMax community while back, I think on Windows what I'm going to do is do a check for the existence of mpv.exe in the working directory and if it does not exist I could use the DirectShow code as a fallback.

**Comment by fundies** [comment]
> @fundies I think you will like this, I have written a video player using DirectShow with some help from someone from the BlitzMax community a while back, I think on Windows what I'm going to do is do a check for the existence of mpv.exe in the working directory and if it does not exist I could use the DirectShow code as a fallback. Trying to compensate for the fact you don't want me doing this. I'm also happy to move the process code into platforms, that way our widgets and other stuff can use it.

You misunderstand, I'm all for a video extension but I just don't want you launching exes to do it. You should be using ffmpeg or whatever to render to a texture directly. Robert or I will likely write an extension to do this eventually but if you want to attempt to do it properly, I will answer what questions I can.

**Comment by time-killer-games** [comment]
@fundies alrighty, I added the direct show code, and even threw in some error checking, (just a little).

**Comment by time-killer-games** [comment]
@RobertBColton @JoshDreamland anything you want me to add to TODO / FIXME list?

**Comment by RobertBColton** [comment]
Sam, we're not too sure about this extension.

The reason is because most video player extensions already do this process API stuff. It's also something that the user is capable of doing in the GML/EDL API already. We don't really want infinite extensions in the default ENIGMA setup. The real meat and potatoes for an official video API extension we would like should hook into our graphics like my original DirectShow extension was planned/going for. In other words things like `video_draw`, `video_get_surface`, and etc so you could do things like draw a video on a 3D cube are what is most important for a game engine to provide.

Not completely blocking it yet but letting you know what the current thoughts are.

**Comment by time-killer-games** [comment]
The video player on Windows uses DirectShow so and it really isn't possible to do this with enigma already unless you were to use mpv media player like on Linux / FreeBSD but I'm not doing that here; only on Linux / FreeBSD. That and there's the fact you can't use video_stop() with execute_shell() at all nor can you with the current DirectShow extension we have which is also broken by comparison. mpv media player is just a cheap work around for Linux and FreeBSD I wrote until I learn how to use libmpv instead which is basically an easier to use ffmpeg wrapper library. The plan was already for that to be temporary.

Wjatever happened to that thing fundies said, "code is better than no code". The way I see it, a fully functional & not-broken video player is far better to have now than to wait **another 10+ years** when ENIGMA might not even exist relevantly anymore.

**Comment by time-killer-games** [comment]
@RobertBColton want to hear another fact about shark poop?

**Comment by RobertBColton** [comment]
We'll see, I suppose that's reasonable under the premise that at the time we have a full proper video extension in ENIGMA, that can draw to 3D cubes and such, you don't complain about us deleting this extension.

**Comment by JoshDreamland** [comment]
It might just be time to add better infrastructure for organizing extensions. We're starting to hit a real problem with the number we have—it's more than a typical person is going to read through and process.

It's true that having some sort of video player code is better than having none, for someone who wants to play video files. But playing video files is a bad idea. Even if we offer perfect support for embedding video files in the game window, we're still doing users a disservice by encouraging them to hack together some kind of recording to play for cutscenes instead of utilizing tools in their game engine to produce one properly in-game.

With users embedding video files instead of doing proper animations, we (and their users) will be storing and transferring hundreds of megabytes of video files for a game that really just needed to move a camera around a few models or send a few sprites across the screen in a specific sequence. This is what timelines are for. Admittedly, we could use better animation sequencing resources. Particularly the kind that let you see what you're doing while you're doing it. Even Yoyo has noticed that one...

The point is, as a game engine, we should be facilitating *good* approaches to *practical* problems. The pushback you see from a lot of us is the result of asking about problems with a particular solution to the actual problem, rather than tackling the actual problem in a careful manner.

In this case, the real problem is "how do I make cool cutscenes in my game." But the problem users are coming to you with (or the question you keep searching out users asking) is "AAARGH HOW I EMBED AVI FILE????//" And then the problems you bring to the table are innumerable (but here's a start):
1. DirectShow is Windows-only and only supports Microsoft codecs
2. There's not an end-all-be-all video player on non-Windows platforms
3. There are window focus and presentation issues when launching an external video player
4. The video player you invoke might already be open with a playlist (paused or playing), which will mean your request is scheduled onto that player instead of actually being carried out
5. The command you fire will return immediately for some players because the launcher spawns a child process

We've now turned one serious problem into five problems not worth solving. This is a waste of energy. Robert has pitched a more tenable (yet still not ideal) approach to playing video files, but the actual problem we should be talking about is "how do I make a great cutscene?"

There are lots of answers in this space! Blender has great animation support. If we go the video approach, we encourage users to use Blender to turn 65KB of Blender files into 300MB of cutscene video. Blender is open-source software... its render library is probably smaller than the video, assuming it's not smaller than ffmpeg.

And there are answers for 2D sequencing, too, but I'm not aware of a great one that's free. Maybe this is a place ENIGMA could shine. I've seen some proofs of concept on YouTube, but none of them have reached a viable product, as far as I know.

So, in a nutshell, more functionality isn't a bad thing, but we want to at least try to guide users toward what the rest of the industry does, rather than encourage them to do the first hack that comes to mind, as terrible as it is for their infrastructure and users.

**Comment by time-killer-games** [comment]
@JoshDreamland the system2 method is supposed to return immediately. The while loop after that call is what makes it blocking.

**Comment by JoshDreamland** [comment]
You haven't arrived at that problem yet. It will make sense when you arrive at it in a few weeks, after you've "fixed" the focus issue.

The point isn't that there are problems. Problems can be solved. The point is that you've picked the solution with the most problems because it resonated with you. We should pick a more tenable solution.

**Comment by time-killer-games** [comment]
@JoshDreamland what Linux distro has that problem? I'd like to install it so i can debug fix that. Not really for this extension but for my process information one for GameMaker which also uses that code. This would be useful for me to fix regardless of whether this extension ever makes it into enigma.

Also, how exactly do you expect me to use this in my game? Should i convert it to PNG and making it a million times bigger in filesize and make it have audio that plays out of sync?
https://www.youtube.com/watch?v=74OBPJUNnqY

**Comment by RobertBColton** [comment]
@time-killer-games The answer to your last question is the same as the answer to the question "How did you render it in the first place?"

**Comment by time-killer-games** [comment]
I just have the video player embedded in the game window using the --wid command line parameter that mpv media player has. DirectShow renders to a window which can also be embedded in the game for the same effect.

**Comment by time-killer-games** [comment]
I misunderstood lol. Yeah i rendered the video with Bryce and enigma doesn't have 3D good enough to look like that.

**Comment by time-killer-games** [comment]
@JoshDreamland I've tested this POSIX code which is also used by my process info extension on MacOS, FreeBSD, and a lot of different Linux distro's. I have yet to see what you were trying to say. I would like to fix it as soon as possible but I have no idea what to do if I can't reproduce it. Sorry for asking so much, I'm just really anxious to get it fixed. *drops mic*

**Comment by time-killer-games** [comment]
> 4\. The video player you invoke might already be open with a playlist (paused or playing), which will mean your request is scheduled onto that player instead of actually being carried out

mpv media player doesn't work that way. It was designed specifically for use with the command line in mind.

**Comment by JoshDreamland** [comment]
There isn't a sequence of fixes that will put the extension in a perfect state. If you try to fix the focus issue by waiting for the video player process to terminate, you'll find that it frequently terminates immediately upon scheduling playback with a separate player process. So your focus fix would actually just hide the video in that case.

In the case of your youtube video, I see a single camera pan motion with a model that its either static or stretching vertically. The camera is zoomed out far enough that you could probably just pull that off with purely 2D sprites, but the obvious solution is to enter 3D mode, draw the lava floor and your stretching model, start the camera object moving downward, and set an alarm for 60 frames to head back to the title screen.

Launch an instance of mpv playing a song and then run your game. If it doesn't decentralize, try VLC.

**Comment by time-killer-games** [comment]
@JoshDreamland alternatively people could just use keyboard_check_direct() to skip the video on linux. Although it would work regardless of the window that's active.

**Comment by RobertBColton** [comment]
The keyboard check direct on Linux is fake, it's not really direct and so that won't work. I did some research before and i'm not even really sure if that function is possible on Linux. Regardless it's never useful for anything, even on Windows where it's implemented correctly. It can miss key presses and releases because it's not message based and works below the hardware interrupt level.

Also ENIGMA needs a command line argument parser for options like the one we use in emake. We should provide default options to all games, one of which would be to skip cutscenes which a lot of games provide a command line option for.

**Comment by time-killer-games** [comment]
@RobertBColton we already have command line parsing. We have parameter_string and parameter_count.

**Comment by RobertBColton** [comment]
That's basic command line parsing, those aren't options like boost options.

**Comment by time-killer-games** [comment]
@RobertBColton MacOS has a Darwin API that did exactly what you are referring to, and since Darwin is open source, and likely uses posix api underneath for that API, we could very easily borrow code from that if it isn't substantial. I've used that api in the past but I'm having trouble finding it atm. Better yet, we could do the same thing with boost, we just need to credit the authors with the appropriate license header in tact. I don't think we should ship the entire boost dependency though, we should only get the relevant code we need since a lot of boost's stuff we are replacing with other things.

**Comment by time-killer-games** [comment]
@JoshDreamland I say we merge it with a disclaimer added to the About.ey to contact samuelvenable@hotmail.com for support and not the enigma development team, because I am the only one who cares to maintain it and provide support such as answer questions like "y it not AVI?" and your other concerns along those lines aren't a relevant burden to any of you if people are coming to only me with the questions and help/feature requests. What say you?

**Comment by fundies** [comment]
Why do you need it to be in the repo. You can just tell users to ```git clone tkg.github/extension ENIGMASystem/Universal/Extensions/videocrap```

**Comment by time-killer-games** [comment]
@fundies also why you guy like and merge Robert's broken and windows-only directshow video player but not mine that isn't broken and working on 3 platforms and therefore is much less limiting?

At the very least we can take out the perfectly working Linux and FreeBSD code for no good reason and replace Robert's broken direct show with my directshow code that doesn't crash and has zero error reporting.

**Comment by fundies** [comment]
> @fundies also why you guy like and merge Robert's broken and windows-only directshow video player but not mine that isn't broken and working on 3 platforms and therefore is much less limiting?
> 
> At the very least we can take out the perfectly working Linux and FreeBSD code for no good reason and replace Robert's broken direct show with my directshow code that doesn't crash and has zero error reporting.
> 
> All of the complaints you have with my windows direct show player are evenly applied to his if not more of an issue with his due to the fact it crashes and doesn't have throw proper errors.
> 
> Seems backwards to me. In every way.

My issue with yours is its just a glorified shell wrapper and not a true extension. However, I do agree some of @RobertBColton's  extensions should be relocated from the main repo as they're in an unusable state and will only serve to confuse / frustrate users trying to use them. In general, I think we should be much stricter with what extensions we want in the main repo as we want to ensure everything shipped with enigma builds and works properly and that eats a lot of CI resources.

**Comment by time-killer-games** [comment]
Ok cool, glad we're on the same page. Either my useable extension gets merged. Or Robert's stuff that isn't useable gets thrown out. Either one would make the situation less troublesome.

**Comment by time-killer-games** [comment]
@fundies the problem with hosting it on my own repository is that hugar is able to install enigma ok on occasion. He has yet to figure out how to get any of my extensions installed from my third party repositories. He has no idea what he's doing and he shows me errors he's magically getting without telling me steps to reproduce. It's a headache I can't avoid any other means. The only reason I wanted this merged was so people like him could actually use it like he wants to.

That and if I were to use this extension at all I don't want to give people install instructions to get my open source games working if they won't ever be able to figure out how to do it on their own without me spoon feeding them every step.

**Comment by fundies** [comment]
> Ok cool, glad we're on the same page to some extent. Either my useable extension gets merged. Or Robert's stuff that isn't useable gets thrown out. Either one would make the situation less troublesome.
> 
> Although, you say it's a glorified shell pooper, what about my directshow code? Did you know that ffmpeg uses that underneath in all likelihood? Either that or they must've transitioned to MS media foundations, but that is still very new and is only compatible with windows 8 and up. I don't know what they'd use on Linux or Mac obviously since those have their own unique API's.

I haven't read the full PR but I would be more keen on merging a fix to Robert's directshow code than this PR that calls shell functions as a poor attempt to support other platforms.

**Comment by fundies** [comment]
> @fundies the problem with hosting it on my own repository is that hugar is able to install enigma ok on occasion. He has yet to figure out how to get any of my extensions installed from my third party repositories. He has no idea what he's doing and he shows me errors he's magically getting without telling me steps to reproduce. It's a headache I can't avoid any other means. The only reason I wanted this merged was so people like him could actually use it like he wants to.
> 
> That and if I were to use this extension at all I don't want to give people install instructions to get my open source games working if they won't ever be able to figure out how to do it on their own without me spoon feeding them every step.

You can just put it in your readme.md on your repo it should really be as simple as running the one command I gave you above. In the future rgm will have some sort of marketplace or whatever to make it easier but for now you'll have to live with running a 1 line command to get your extension.

**Comment by time-killer-games** [comment]
Robert's pr has a lot of problems that don't exist with this one. 

The video waits until playback is done before any window interaction can be made. Mine runs in it's own thread and resizes dynamically to fill the client area to solve that issue. That is the default behavior but a video_set_rectangle function I could easily add to override that. I could also remove all the shell stuff as none of it is really needed on Windows and Linux/FreeBSD can just be removed if you abosutely insist. I can also get rid of that second argument to set the window to embed in and make it just do window_handle() hard-coded.

**Comment by time-killer-games** [comment]
> In the future rgm will have some sort of marketplace or whatever to make it easier but for now you'll have to live with running a 1 line command to get your extension.

yeah I've written "easy install" scripts that download my extensions with base enigma. Didn't really help him at all. I've ran the scripts to do fresh installs that include my extensions and have done it for a long ass time now so I have no idea what hugar thinks he's doing.

**Comment by fundies** [comment]
> Robert's pr has a lot of problems that don't exist with this one.
> 
> The video waits until playback is done before any window interaction can be made. Mine runs in it's own thread and resizes dynamically to fill the client area to solve that issue. That is the default behavior but a video_set_rectangle function I could easily add to override that. I could also remove all the shell stuff as none of it is really needed on Windows and Linux/FreeBSD can just be removed if you abosutely insist. I can also get rid of that second argument to set the window to embed in and make it just do window_handle() hard-coded.
> 
> I'm perfectly fine with improving this pr, I just don't want to scratch it cold turkey when I find that counterproductive.

By "fixing" Robert's code you could replace it entirely with your direct show code for all I care my main issue is you using the shell instead of a c++ APIs to play videos on platforms that don't support directshow when you should really be using ffmpeg to do that.

**Comment by time-killer-games** [comment]
Sounds fine to me, I really didn't mean to make such a huge fuss. I think we all want what's best for the project we just have different ideas of what that looks like, and unfortunately you guys disagree with me a lot more on these things than you seem to with each other lol I wish I just had ideas you guys liked more often and problem solved.

**Comment by fundies** [comment]
> Sounds fine to me, I really didn't mean to make such a huge fuss. I think we all want what's best for the project we just have different ideas of what that looks like, and unfortunately you guys disagree with me a lot more on these things than you seem to with each other lol I wish I just had ideas you guys liked more often and problem solved.

Robert josh and I all disagree on things too. This is why I try to discuss a proposal before wasting my time writing it only to have josh or Robert shit on it later and often they have ways to do it better or more easily too.

**Comment by time-killer-games** [comment]
I think I'm going to make all the changes we talked about locally and then just upload what i have to a new pr. I'll keep this one opened for people to be able to see the branch in the meantime. But once I open the new one I'll go ahead and close this.

**Comment by time-killer-games** [comment]
@fundies Alright, this no longer uses command line crapola, and it also no longer uses direct show. It uses libmpv now, which is based on ffmpeg, and I added the ability to pause / unpause.

@JoshDreamland if my impression is correct, if the game doesn't static link the limpv library and isn't a derivative work of the original, (due the LGPL) is there anything i should be concerned about regarding distributing those dll's (do they have to provide a license file, the source code to the dll's etc or not?)

**Comment by time-killer-games** [comment]
@fundies turns out libmpv can do exactly what you want with rendering to a texture: https://www.youtube.com/watch?v=dA8J241K9dw

File: ENIGMAsystem/SHELL/Universal_System/Extensions/VideoPlayer/videoplayer.cpp
Comment (JoshDreamland) [inline ENIGMAsystem/SHELL/Universal_System/Extensions/VideoPlayer/videoplayer.cpp:44]: Instead of all these maps, it's safer to use a single map of ID to a struct:
```cpp
struct VideoData {
  mpv_handle *mpv;
  std::string window_id;
  std::string gpu_api;
  int volume;
  bool is_playing;
  bool is_fullscreen;
};
static std::map<std::string, std::string> videos;
```

File: ENIGMAsystem/SHELL/Universal_System/Extensions/VideoPlayer/videoplayer.cpp
Comment (JoshDreamland) [inline ENIGMAsystem/SHELL/Universal_System/Extensions/VideoPlayer/videoplayer.cpp:71]: Using a struct allows you to return internal IDs for these videos rather than converting between pointers and strings. It's not technically valid to convert a pointer to an integer type, or vice-versa, and much less to use a string implementation (which relies on the resulting string capturing the full precision of pointer values).

File: ENIGMAsystem/SHELL/Universal_System/Extensions/VideoPlayer/videoplayer.cpp
Comment (JoshDreamland) [inline ENIGMAsystem/SHELL/Universal_System/Extensions/VideoPlayer/videoplayer.cpp:82]: At the point where you're comparing a string to "yes"/"no", you should be storing a boolean and replacing it with (bool_value ? "yes" : "no") in the places where a string is required

**Review by JoshDreamland — COMMENTED** [review]
This seems like a much simpler, saner approach. I like this option. A couple of points concern me (the most important is serializing pointers as strings; that's two kinds of not safe).

File: ENIGMAsystem/SHELL/Universal_System/Extensions/VideoPlayer/videoplayer.cpp
Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Universal_System/Extensions/VideoPlayer/videoplayer.cpp:71]: Actually it technically is valid, and the people who created libmpv encourage doing this for the window id. Not only that, Microsoft Visual Basic handles all pointers as Int64 under the hood. So if MS does it, I can't imagine you could be right about this. I'm going to change it regardless because I know you won't merge it without me doing that.

**Comment by time-killer-games** [comment]
Thanks josh for the code review! I'm addressing your critiques right now, finally. Should be done applying changes shortly.

File: ENIGMAsystem/SHELL/Universal_System/Extensions/VideoPlayer/videoplayer.cpp
Comment (JoshDreamland) [reply ENIGMAsystem/SHELL/Universal_System/Extensions/VideoPlayer/videoplayer.cpp:71]: Here is some reading for you: https://wiki.sei.cmu.edu/confluence/display/c/INT36-C.+Converting+a+pointer+to+integer+or+integer+to+pointer

Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Universal_System/Extensions/VideoPlayer/videoplayer.cpp:71]: I did figure there were exceptions where what you say is true, and all that article does is confirm we are both partially right. It isn't all or nothing. There are implementations of pointers where that is safe, other times when it isn't. This has been tested like hogweed, and I can assure you has no side effects. But like I said, I'm applying your changes now.

**Comment by time-killer-games** [comment]
Thanks for the link for explanation. That gave a lot of insight to a real issue here. Anyway, wish granted. Let me know if you need anything else before this is merged. And remember, I am happy to expand this extension to support rendering to a texture. That was already a part of my plans because I have someone paying me to do this same thing but for a different game engine (not even gamemaker).

**Comment by time-killer-games** [comment]
@JoshDreamland alright, I tested everything locally and it's working great. Please note I implemented GM 8.1's splash video functions since the last time you saw this pr.

File: ENIGMAsystem/SHELL/Universal_System/Extensions/VideoPlayer/videoplayer.cpp
Comment (time-killer-games) [inline ENIGMAsystem/SHELL/Universal_System/Extensions/VideoPlayer/videoplayer.cpp:214]: @fundies i know how much you hate these platform macros so please don't worry I already intend on removing them, I wanted to tell you in advance I'm going to separate this into a windows and mac helper and a non-windows/mac function stub in the Makefile because it'll be a while before i can do it. I need to discuss with Josh something first in regards to the OS X (posix) equivalent to Win32 PeekMessage. Linux doesn't need the posix equivalent to peekmessage for this code to work but windows and mac will have the program not respond and crash without a message pump of some kind.

File: ENIGMAsystem/SHELL/Universal_System/Extensions/VideoPlayer/videoplayer.cpp
Comment (time-killer-games) [inline ENIGMAsystem/SHELL/Universal_System/Extensions/VideoPlayer/videoplayer.cpp:220]: Note for future maintainers: I need this code to process mouse clicks on the given window id instead of mouse_check_button because the window id might not be the game window but a different one if specified by the developer, so anyway, that's why I'm doing it this way. Linux doesn't need this code because there is no message pump blocking the input.conf file from processing input.

**Review by JoshDreamland — APPROVED** [review]
This looks good. Is the CI busted again, or does this change somehow break it? afaict it should not...

**Comment by time-killer-games** [comment]
@JoshDreamland I believe the CI is just busted. I guess we can wait a few days and retry the tests then and see if that makes a difference.

File: ENIGMAsystem/SHELL/Universal_System/Extensions/VideoPlayer/videoplayer.cpp
Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Universal_System/Extensions/VideoPlayer/videoplayer.cpp:269]: Yeah i didn't realize that until after the fact. I've done that before I just didn't think of it this time for whatever reason.

**Comment by time-killer-games** [comment]
@JoshDreamland just leaving the note that splash_set_window(window_identifier()) and/or video_set_window_identifier(window_identifier()) should be called before attempting to embed in the game window. I made the default to be no window in case the developer wants to play audio files instead or embed the video in a different window that isn't necessarily the game. Let me know if you want to make the game window the default instead. I wasn't sure what your preference was.

File: ENIGMAsystem/SHELL/Universal_System/Extensions/VideoPlayer/Makefile
Comment (time-killer-games) [inline ENIGMAsystem/SHELL/Universal_System/Extensions/VideoPlayer/Makefile:13]: Well, I have absolutely no idea why azure is getting undefined references but appveyor is not. They both should be working, looking at this Makefile.

Comment (fundies) [reply ENIGMAsystem/SHELL/Universal_System/Extensions/VideoPlayer/Makefile:13]: its because the linux pkg config is getting called you need to use $(pkg-config) var so that  it calls the correct one.

Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Universal_System/Extensions/VideoPlayer/Makefile:13]: Could you give an example of this that already exists somewhere in the engine that I could use as a reference?

Comment (fundies) [reply ENIGMAsystem/SHELL/Universal_System/Extensions/VideoPlayer/Makefile:13]: I don't remember where. I think you just put $(PKG-CONFIG) in the makefile where you have pkg-config it may or may not need to be all caps

Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Universal_System/Extensions/VideoPlayer/Makefile:13]: I discovered a bug with Mac. The video end playback event doesn't get triggered at the end of video playback until after you press the keyboard or mouse. I'm just going to simulate a mouse button on the window when current position equals duration because I don't care enough to go with a more correct solution. If you or Josh feel differently I act accordingly I just don't want to do it unless I have to. When that's done, I'll check out what you said.

Comment (time-killer-games) [reply ENIGMAsystem/SHELL/Universal_System/Extensions/VideoPlayer/Makefile:13]: Nevermind, I figured out the right way to do it that doesn't involve simulating a mouse press.

**Comment by time-killer-games** [comment]
@JoshDreamland halp

**Comment by time-killer-games** [comment]
@fundies I tried both $(PKG-CONFIG) and lower case both give me big doo doo

**Comment by time-killer-games** [comment]
@JoshDreamland libmpv autodetects when youtube-dl is installed in the case where the user passed a supported URL to video_add() and so i added youtube-dl as a dependency to Travis and Appveyor so when the time comes that I add tests, there can be a youtube URL video test as well. The end user does not need youtube-dl installed unless they want to load URL's, otherwise passing a URL to libmpv when youtube-dl isn't installed it will not prevent the EXE from running, it will just fail silently.

**Pull request closed without merging**
